src/analytics.rs

```diff
@@ -0,0 +1,194 @@
+use crate::amount::AsSats;
+use crate::async_runtime::Handle;
+use crate::errors::Result;
+use crate::locker::Locker;
+use crate::util::unix_timestamp_to_system_time;
+use crate::{derive_key_pair_hex, ExchangeRate, InvoiceDetails, UserPreferences};
+use breez_sdk_core::{
+    InvoicePaidDetails, Payment, PaymentDetails, PaymentFailedData, ReceivePaymentResponse,
+};
+use log::{info, warn};
+use parrot::{AnalyticsClient, AnalyticsEvent, PayFailureReason, PaymentSource};
+use std::sync::{Arc, Mutex};
+use std::time::SystemTime;
+use uuid::Uuid;
+
+const BACKEND_ANALYTICS_DERIVATION_PATH: &str = "m/82640931'/0'/0";
+
+pub(crate) struct AnalyticsInterceptor {
+    pub analytics_client: Arc<AnalyticsClient>,
+    pub user_preferences: Arc<Mutex<UserPreferences>>,
+    pub rt_handle: Handle,
+}
+
+/// Includes metadata about a payment the client received in anyway.
+pub struct PaymentMetadata {
+    pub source: PaymentSource,
+    /// The current time the client started the pay process (e.g. scanned the invoice)
+    pub process_started_at: SystemTime,
+}
+
+/// Includes metadata about a invoice the user created.
+pub struct InvoiceCreationMetadata {
+    /// The currency the user used to define the requested amount e.g. chf/sat
+    pub request_currency: String,
+}
+
+impl AnalyticsInterceptor {
+    pub fn new(
+        analytics_client: Arc<AnalyticsClient>,
+        user_preferences: Arc<Mutex<UserPreferences>>,
+        rt_handle: Handle,
+    ) -> Self {
+        Self {
+            analytics_client,
+            user_preferences,
+            rt_handle,
+        }
+    }
+
+    pub fn pay_initiated(
+        &self,
+        invoice_details: InvoiceDetails,
+        metadata: PaymentMetadata,
+        paid_amount: Option<u64>,
+        exchange_rate: Option<ExchangeRate>,
+    ) {
+        let invoice_amount = invoice_details.amount.map(|a| a.sats.as_sats().msats);
+        let paid_amount_msat = match paid_amount {
+            None => invoice_amount.unwrap(),
+            Some(a) => a,
+        };
+
+        let user_currency = self.user_preferences.lock_unwrap().fiat_currency.clone();
+        let analytics_client = self.analytics_client.clone();
+
+        self.rt_handle.spawn(async move {
+            log_error(
```

Comment:
This is a suggestion to make it clear that we are ignoring the error. Feel free to ignore it if you don't like it.
```suggestion
            log_ignore_error(
```

Reply:
Another improvement would be to implement a method on `Result` such that we call `analytics_client
                    .report_event(...).await.log_ignore_error()`